### PR TITLE
CMS Signing support

### PR DESCRIPTION
**Pull request by Lukasa**
This patch adds support for basic CMS operations. We aren't sure that the API surface here is right, so for now we're going to provide this under an SPI guard so that we can make any amendments we might need.

### CODE REVIEW

File: Sources/X509/CryptographicMessageSyntax/CMSOperations.swift [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+            // This force-unwrap is safe: we know there are certificates because we've located at least one certificate from this set!
+            let untrustedIntermediates = CertificateStore(signedData.certificates!)
+            var verifier = Verifier(rootCertificates: trustRoots, policy: policy)
+            let result = await verifier.validate(leafCertificate: signingCert, intermediates: untrustedIntermediates)
+
+            switch result {
+            case .validCertificate:
+                return .validSignature(.init(signer: signingCert))
+            case .couldNotValidate(let validationFailures):
+                return .unableToValidateSigner(.init(validationFailures: validationFailures))
```
Comment (yim-lee) [inline Sources/X509/CryptographicMessageSyntax/CMSOperations.swift:146]: Can we also include `signingCert` in `unableToValidateSigner`?

Comment (Lukasa) [reply Sources/X509/CryptographicMessageSyntax/CMSOperations.swift]: Done.

File: Sources/X509/CryptographicMessageSyntax/CMSOperations.swift [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+    @_spi(CMS)
+    public enum Error: Swift.Error {
+        case incorrectCMSVersionUsed
+    }
+
+    @_spi(CMS)
+    public enum SignatureVerificationResult {
+        case validSignature(Valid)
+        case unableToValidateSigner(SignerValidationFailure)
+        case invalidCMSBlock(InvalidCMSBlock)
```
Comment (dnadoba) [inline Sources/X509/CryptographicMessageSyntax/CMSOperations.swift:177]: This looks like a good use case for a typed error combined with `Result` e.g.:
```suggestion
    public typealias SignatureVerificationResult = Result<Valid, VerificationError>
    public enum VerificationError: Error {
        case unableToValidateSigner(SignerValidationFailure)
        case invalidCMSBlock(InvalidCMSBlock)
```

File: Sources/X509/CryptographicMessageSyntax/CMSOperations.swift [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+    func certificate(signerInfo: CMSSignerInfo) throws -> Certificate? {
+        switch signerInfo.signerIdentifier {
+        case .issuerAndSerialNumber(let issuerAndSerialNumber):
+            for cert in self {
+                if cert.issuer == issuerAndSerialNumber.issuer && cert.serialNumber == issuerAndSerialNumber.serialNumber {
+                    return cert
+                }
+            }
+        case .subjectKeyIdentifier:
+            // This is unsupported for now.
```
Comment (dnadoba) [inline Sources/X509/CryptographicMessageSyntax/CMSOperations.swift:227]: Should we throw instead for now?

File: Sources/X509/CryptographicMessageSyntax/CMSSignerInfo.swift [hunk 4]
```diff
@@ -33,20 +33,25 @@ import SwiftASN1
...
-    var signatureAlgorithm: AlgorithmIdentifier
-    var signature: ASN1OctetString
-    
+    @usableFromInline var version: CMSVersion
+    @usableFromInline var signerIdentifier: CMSSignerIdentifier
+    @usableFromInline var digestAlgorithm: AlgorithmIdentifier
+    @usableFromInline var signatureAlgorithm: AlgorithmIdentifier
+    @usableFromInline var signature: ASN1OctetString
+
+    @inlinable
```
Comment (dnadoba) [inline Sources/X509/CryptographicMessageSyntax/CMSSignerInfo.swift:54]: Do we really need `@inlinable` here or is `@usableFromInline` enough?  AFAICT nothing is generic and therefore doesn't really benefit from `@inlinable`

Comment (Lukasa) [reply Sources/X509/CryptographicMessageSyntax/CMSSignerInfo.swift:54]: Having everything be `@inlinable` is still useful, even without generics: it allows the compiler to observe that these functions don't affect global state, causing ARC traffic to be reduced.

File: Sources/X509/CryptographicMessageSyntax/CMSOperations.swift (hunk 3 above)
Comment (Lukasa) [reply Sources/X509/CryptographicMessageSyntax/CMSOperations.swift:227]: We will, because we'll fail to find a signer.

**Review by dnadoba — APPROVED** [review]
Great! LGTM now

File: Sources/X509/CryptographicMessageSyntax/CMSOperations.swift (hunk 1 above)
Comment (yim-lee) [reply Sources/X509/CryptographicMessageSyntax/CMSOperations.swift]: 🙏

**Pull request merged**